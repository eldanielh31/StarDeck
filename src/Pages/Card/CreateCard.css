.container {
    display: flex;
    flex-direction: column;
    height: 100vh;
}

.content {
    display: flex;
    flex: 1;
    padding: 20px;
}

.containerCard {
    padding-right: 20px;
    border-right: 5px solid #000;
    display: flex;
    justify-content: center;
    align-items: center;
    width: 45%;
}

.containerForm {
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
    width: 55%;
}

.wrapperForm {
    background-color: whitesmoke;
    padding: 1em 5em 1em 5em;
    box-shadow: 10px 10px 20px;
    border-radius: 15px;
}

.containerForm h1 {
    font-size: 50px;
    margin: 0 0 1em;
}

.card-content {
    color: #000;
    width: 450px;
    border: 13px solid transparent;
    box-shadow: 10px 10px 20px;
    border-radius: 10px;
}

.card-content h3,
h4 {
    background-color: #f5b267;
    border: 3px solid;
    border-radius: 20px/50px;
    box-shadow: 0 0 10px;
    margin: 0 0 1px;
    padding: 0 0 2px 7px;
}

.image-box {
    border: 3px solid;
    box-shadow: 0 0 10px;
    display: block;
    margin: auto;
    padding: 1px;
    max-width: 400px;
    max-height: 400px;
}

.info-box {
    background-color: whitesmoke;
    border: 3px solid;
    border-radius: 5px;
    box-shadow: 0 0 10px;
    padding: 7px 7px 0 7px;
    text-align: justify;
}

.list {
    list-style-type: none;
    margin: 0;
    padding: 0;
}

.list-item-title {
    font-weight: bold;
}

.description {
    font-size: 15px;
}

.credits {
    color: #001917;
    font-size: 15px;
    margin: 0 7px 10px 0;
}

.containerInputs {
    display: flex;
    flex-direction: column;
    margin-bottom: 1em;

}

.inputCard {
    font-size: 20px;
}

.containerInputs input {
    background-color: whitesmoke;
    height: 30px;
    font-size: 15px;

    border: none;
    border-bottom: 1px solid black;
    outline: none;
}

.containerInputs select {
    height: 30px;
    font-size: 15px;
    border: none;
    border-bottom: 1px solid black;
    outline: none;
    background-color: whitesmoke;
}

.containerInputs textarea {
    min-height: 100px;
    min-width: 400px;
    font-size: 15px;
    border-radius: 10px;
    border: #001917 solid 2px;
    max-width: 600px;
    max-height: 150px;
}

.button-54 {
    margin-top: 1.5em;
    font-size: 16px;
    letter-spacing: 2px;
    text-decoration: none;
    text-transform: uppercase;
    color: #000;
    cursor: pointer;
    border: 3px solid;
    box-shadow: 1px 1px 0px 0px, 2px 2px 0px 0px, 3px 3px 0px 0px, 4px 4px 0px 0px, 5px 5px 0px 0px;
    position: relative;
    user-select: none;
    -webkit-user-select: none;
    touch-action: manipulation;
}

.button-54:active {
    box-shadow: 0px 0px 0px 0px;
    top: 5px;
    left: 5px;
}

@media (min-width: 768px) {
    .button-54 {
        padding: 0.25em 0.75em;
    }
}