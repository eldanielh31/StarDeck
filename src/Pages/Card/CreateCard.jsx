import React, { useRef } from 'react'
import "./CreateCard.css"
import { useState } from 'react';
import { v4 as uuidv4 } from "uuid";
import * as htmlToImage from "html-to-image";
import { createCarta } from '../../requestApi';

/* The above code is a React component that creates a form for users to create a card. The form
includes input fields for the card's name, image, description, type, race, cost, and energy. The
component also includes a dropdown menu for selecting the card's rarity and race. When the user
submits the form, a unique code is generated, the form data is logged, and a PNG image of the card
element is downloaded. The component uses useState and useRef hooks to manage state and reference
the card element. */
function CreateCard() {

    const elementCard = useRef(null);

    /* `const types` is an array of objects that contains different types of card rarity with their
    corresponding labels. This array is used to populate a dropdown menu in the form where the user
    can select the type of the card they are creating. */
    const types = [
        { value: 'ultra-rara', label: 'Ultra-Rara', color: ' #f1c40f' },
        { value: 'muy-rara', label: 'Muy Rara', color: ' #b353be ' },
        { value: 'rara', label: 'Rara', color: '  #294ef7 ' },
        { value: 'normal', label: 'Normal', color: ' #2a8121 ' },
        { value: 'basica', label: 'Básica', color: ' #ed5b41 ' },
    ]

    const races = [
        { value: 'elfo', label: 'Elfo' },
        { value: 'paladin', label: 'Paladin' },
        { value: 'barbaro', label: 'Barbaro' },
    ]

    const [name, setName] = useState('Nombre de Carta');
    const [type, setType] = useState('Tipo');
    const [color, setColor] = useState(' #ed5b41 ')
    const [cardImg, setCardImg] = useState('https://static.vecteezy.com/system/resources/previews/004/141/669/non_2x/no-photo-or-blank-image-icon-loading-images-or-missing-image-mark-image-not-available-or-image-coming-soon-sign-simple-nature-silhouette-in-frame-isolated-illustration-vector.jpg');
    const [description, setDescription] = useState('Descripcion de la carta');
    const [race, setRace] = useState('...');
    const [cost, setCost] = useState('...');
    const [energy, setEnergy] = useState('...');
    const [code, setCode] = useState('C-XXXXXXXXXX');

    /**
     * The function handles form submission, generates a unique code, logs form data and downloads a
     * PNG image of a card element.
     */
    const handleSubmit = async (e) => {
        e.preventDefault();
        const codigo = 'C-' + uuidv4().split('').reverse().join('').slice(0, 12);

        setCode(codigo);

        await htmlToImage.toPng(elementCard.current)
            .then((dataUrl) => {
                const link = document.createElement('a');
                link.download = `${codigo}.png`;
                link.href = dataUrl;
                link.click();

                let carta = {
                    nombre: name,
                    tipo: type,
                    imagen: dataUrl,
                    descripcion: description,
                    raza: race,
                    costo: cost,
                    energia: energy,
                    estado: true,
                    id: codigo
                }

                createCarta(carta).then(
                    () => {
                        console.log("Enviado correcto.");
                    }
                );

            });
        
        

    }

    const handleImageChange = (file) => {
        if (!file) {
            setCardImg('');
            return;
        }

        fileToDataUri(file)
            .then(dataUri => {
                setCardImg(dataUri)
            })
    }

    const fileToDataUri = (file) => new Promise((resolve, reject) => {
        const reader = new FileReader();
        reader.onload = (event) => {
            resolve(event.target.result)
        };
        reader.readAsDataURL(file);
    })

    return (
        <div className="container">
            <div className="content">
                <div className="containerCard">
                    <div className="card-content" style={{ backgroundColor: color }} ref={elementCard}>
                        <h3>{name}</h3>
                        <img className="image-box" src={cardImg} alt="imagen de la carta" />
                        <h4>{type}</h4>
                        <div className="info-box">
                            <ul className="list">
                                <li><span className="list-item-title">Raza: </span>{race}</li>
                                <li><span className="list-item-title">Costo: </span>{cost}</li>
                                <li><span className="list-item-title">Energy: </span>{energy}</li>
                            </ul>
                            <p className="description">{description}</p>
                            <p className="credits" align="right">{code}</p>
                        </div>
                    </div>

                </div>
                <div className="containerForm">
                    <div className='wrapperForm'>
                        <h1>Crear carta</h1>
                        <div className='containerFrom'>
                            <form onSubmit={handleSubmit}>
                                <div className='containerInputs'>
                                    <label htmlFor="name" className='inputCard'>Nombre:</label>
                                    <input type="text" id="name" minLength={5} maxLength={30} onChange={(e) => setName(e.target.value)} required />
                                </div>
                                <div className='containerInputs'>
                                    <label htmlFor="image" className='inputCard'>Imagen:</label>
                                    <input type="file" id="image" onChange={(e) => handleImageChange(e.target.files[0] || null)} required />
                                </div>
                                <div className='containerInputs'>
                                    <label htmlFor="description" className='inputCard'>Descripción:</label>
                                    <textarea id="description" maxLength={1000} onChange={(e) => setDescription(e.target.value)} required />
                                </div>
                                <div className='containerInputs'>
                                    <label htmlFor="description" className='inputCard'>Tipo:</label>
                                    <select id="type" name="type" onChange={(event) => setType(event.target.value) } required>
                                        <option value="">Seleccione un tipo</option>
                                        {types.map((opcion) => (
                                            <option key={opcion.value} onClick={(e) => setColor(opcion.color)} value={opcion.value}>{opcion.label}</option>
                                        ))}
                                    </select>
                                </div>
                                <div className='containerInputs'>
                                    <label htmlFor="race" className='inputCard'>Raza:</label>
                                    <select id="race" name="race" onChange={(event) => setRace(event.target.value)} required>
                                        <option value="">Seleccione una raza</option>
                                        {races.map((opcion) => (
                                            <option key={opcion.value} value={opcion.value}>{opcion.label}</option>
                                        ))}
                                    </select>
                                </div>
                                <div className='containerInputs'>
                                    <label htmlFor="cost" className='inputCard'>Costo:</label>
                                    <input id="number" type='number' min={0} max={100} onChange={(e) => setCost(e.target.value)} required />
                                </div>
                                <div className='containerInputs'>
                                    <label htmlFor="energy" className='inputCard'>Energia:</label>
                                    <input id="number" type='number' min={-100} max={100} onChange={(e) => setEnergy(e.target.value)} required />
                                </div>

                                <button className='button-54' type="submit">Crear carta</button>
                            </form>
                        </div>
                    </div>

                </div>
            </div>
        </div>
    )
}

export default CreateCard