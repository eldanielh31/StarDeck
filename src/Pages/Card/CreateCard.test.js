import React from 'react';
import CreateCard from './CreateCard.jsx';
import { render } from '@testing-library/react';

describe('Pantalla Cartas', () => {

  // Prueba de renderizado
  it('Renderiza correctamente', () => {
    const { container } = render(<CreateCard />);
    expect(container.firstChild).toMatchSnapshot();
  });

  // Prueba de propiedades
  it('Muestra el texto adecuado', () => {
    const { getByText, getAllByText } = render(<CreateCard/>);
    expect(getAllByText('Nombre:')[0]).toBeInTheDocument();
    expect(getAllByText('Imagen:')[0]).toBeInTheDocument();
    expect(getAllByText('Descripci√≥n:')[0]).toBeInTheDocument();
    expect(getAllByText('Tipo:')[0]).toBeInTheDocument();
    expect(getAllByText('Raza:')[0]).toBeInTheDocument();
    expect(getAllByText('Costo:')[0]).toBeInTheDocument();
    expect(getAllByText('Energia:')[0]).toBeInTheDocument();
    
  });

});

