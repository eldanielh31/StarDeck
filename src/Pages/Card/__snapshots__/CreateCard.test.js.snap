// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pantalla Cartas Renderiza correctamente 1`] = `
<div
  class="container"
>
  <div
    class="content"
  >
    <div
      class="containerCard"
    >
      <div
        class="card-content"
        style="background-color: rgb(237, 91, 65);"
      >
        <h3>
          Nombre de Carta
        </h3>
        <img
          alt="imagen de la carta"
          class="image-box"
          src="https://static.vecteezy.com/system/resources/previews/004/141/669/non_2x/no-photo-or-blank-image-icon-loading-images-or-missing-image-mark-image-not-available-or-image-coming-soon-sign-simple-nature-silhouette-in-frame-isolated-illustration-vector.jpg"
        />
        <h4>
          Tipo
        </h4>
        <div
          class="info-box"
        >
          <ul
            class="list"
          >
            <li>
              <span
                class="list-item-title"
              >
                Raza: 
              </span>
              ...
            </li>
            <li>
              <span
                class="list-item-title"
              >
                Costo: 
              </span>
              ...
            </li>
            <li>
              <span
                class="list-item-title"
              >
                Energy: 
              </span>
              ...
            </li>
          </ul>
          <p
            class="description"
          >
            Descripcion de la carta
          </p>
          <p
            align="right"
            class="credits"
          >
            C-XXXXXXXXXX
          </p>
        </div>
      </div>
    </div>
    <div
      class="containerForm"
    >
      <div
        class="wrapperForm"
      >
        <h1>
          Crear carta
        </h1>
        <div
          class="containerFrom"
        >
          <form>
            <div
              class="containerInputs"
            >
              <label
                class="inputCard"
                for="name"
              >
                Nombre:
              </label>
              <input
                id="name"
                maxlength="30"
                minlength="5"
                required=""
                type="text"
              />
            </div>
            <div
              class="containerInputs"
            >
              <label
                class="inputCard"
                for="image"
              >
                Imagen:
              </label>
              <input
                id="image"
                required=""
                type="file"
              />
            </div>
            <div
              class="containerInputs"
            >
              <label
                class="inputCard"
                for="description"
              >
                Descripción:
              </label>
              <textarea
                id="description"
                maxlength="1000"
                required=""
              />
            </div>
            <div
              class="containerInputs"
            >
              <label
                class="inputCard"
                for="description"
              >
                Tipo:
              </label>
              <select
                id="type"
                name="type"
                required=""
              >
                <option
                  value=""
                >
                  Seleccione un tipo
                </option>
                <option
                  value="ultra-rara"
                >
                  Ultra-Rara
                </option>
                <option
                  value="muy-rara"
                >
                  Muy Rara
                </option>
                <option
                  value="rara"
                >
                  Rara
                </option>
                <option
                  value="normal"
                >
                  Normal
                </option>
                <option
                  value="basica"
                >
                  Básica
                </option>
              </select>
            </div>
            <div
              class="containerInputs"
            >
              <label
                class="inputCard"
                for="race"
              >
                Raza:
              </label>
              <select
                id="race"
                name="race"
                required=""
              >
                <option
                  value=""
                >
                  Seleccione una raza
                </option>
                <option
                  value="elfo"
                >
                  Elfo
                </option>
                <option
                  value="paladin"
                >
                  Paladin
                </option>
                <option
                  value="barbaro"
                >
                  Barbaro
                </option>
              </select>
            </div>
            <div
              class="containerInputs"
            >
              <label
                class="inputCard"
                for="cost"
              >
                Costo:
              </label>
              <input
                id="number"
                max="100"
                min="0"
                required=""
                type="number"
              />
            </div>
            <div
              class="containerInputs"
            >
              <label
                class="inputCard"
                for="energy"
              >
                Energia:
              </label>
              <input
                id="number"
                max="100"
                min="-100"
                required=""
                type="number"
              />
            </div>
            <button
              class="button-54"
              type="submit"
            >
              Crear carta
            </button>
          </form>
        </div>
      </div>
    </div>
  </div>
</div>
`;
